### PR TITLE
CB-10399 Added Appium tests

### PR DESCRIPTION
**Pull request by alsorokin**
https://issues.apache.org/jira/browse/CB-10399

### CODE REVIEW

**Comment by nikhilkh** [comment]
@dblotsky @riknoll to review.

File: appium-tests/ios/ios.spec.js [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+                    .execute('document.getElementById("info").innerHTML = "' + STARTING_MESSAGE + '";')
+                    .finally(done);
+            }
+            done();
+        }, 3 * MINUTE);
+
+        afterEach(function (done) {
+            if (!errorFlag || stopFlag) {
+                // either there's no error or we've failed irrecoverably
+                // nothing to worry about!
```
Comment (dblotsky) [inline appium-tests/ios/ios.spec.js]: Is there a way to loudly announce which of those two happened?

File: appium-tests/ios/ios.spec.js [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+                }
+                console.log('Webview context is ' + webviewContext);
+                done();
+            }, function onError (err) {
+                console.log(err);
+            })
+            .finally(done);
+    }, MINUTE);
+
+    describe('specs', function () {
```
Comment (dblotsky) [inline appium-tests/ios/ios.spec.js]: Which specs?

File: appium-tests/android/android.spec.js [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    // this indicates that there was a critical error and we should try to recover:
+    var errorFlag = false;
+    // this indicates that we couldn't restore Appium session and should fail fast:
+    var stopFlag = false;
+
+    function win() {
+        expect(true).toBe(true);
+    }
+
+    function fail(error) {
```
Comment (dblotsky) [inline appium-tests/android/android.spec.js]: Are we using a version of Jasmine that has a `fail()` function?

File: appium-tests/android/android.spec.js [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+                        // we should try to recreate the session for the following tests
+                        if (msg.indexOf('Error response status: 6') >= 0 ||
+                            msg.indexOf('Error response status: 7') >= 0) {
+                            errorFlag = true;
+                        }
+                        return result;
+                    }
+                };
+            },
+            toContain: function () {
```
Comment (dblotsky) [inline appium-tests/android/android.spec.js]: There is a `toContain` matcher in [Jasmine's standard matchers][matchers].

[matchers]: http://jasmine.github.io/edge/introduction.html#section-Included_Matchers

File: appium-tests/android/android.spec.js [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+
+    it('camera.ui.util configuring driver and starting a session', function (done) {
+        stopFlag = true; // just in case of timeout
+        getDriver().then(function () {
+            stopFlag = false;
+        }, function (error) {
+            return fail(error);
+        })
+        .finally(done);
+    }, 5 * MINUTE);
```
Comment (dblotsky) [inline appium-tests/android/android.spec.js]: Is there an expect in this spec?

File: appium-tests/android/android.spec.js [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+        afterEach(function (done) {
+            if (!errorFlag || stopFlag) {
+                // either there's no error or we've failed irrecoverably
+                // nothing to worry about!
+                done();
+                return;
+            }
+            // recreate the session if there was a critical error in a previous spec
+            stopFlag = true; // we're going to set this to false if we're able to restore the session
```
Comment (dblotsky) [inline appium-tests/android/android.spec.js]: Please place comments on their own line.

File: appium-tests/android/android.spec.js [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+                    expect(html).toContain('"displayName": null');
+                })
+                .then(win, fail)
+                .finally(done);
+        }, 5* MINUTE);
+
+        it('contacts.ui.spec.4 Special characters and escape sequences', function (done) {
+            var contactName = contactsHelper.getContactName('Н€йромонах', 'ФеофаЊ');
+            enterTest()
+                .execute(contactsHelper.getAddContactCode(contactName.formatted, contactName, '4 5 6789'))
```
Comment (dblotsky) [inline appium-tests/android/android.spec.js]: I don't know if this string in particular occurs in other places, but please check if any literals like it can be factored out into constants.

File: appium-tests/helpers/contactsHelper.js [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+*/
+
+'use strict';
+
+function prepare(item) {
```
Comment (dblotsky) [inline appium-tests/helpers/contactsHelper.js]: Please give this function a more descriptive name.

File: appium-tests/helpers/contactsHelper.js [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        '            removeNext(nextToRemove);\n' +
+        '        });\n' +
+        '    }\n' +
+        '    removeNext(nextToRemove);\n' +
+        '}, function(e) {\n' +
+        '   results.innerHTML = "Search failed: error " + e.code;\n' +
+        '}, obj);';
+
+    return result;
+};
```
Comment (dblotsky) [inline appium-tests/helpers/contactsHelper.js]: Why is there code as strings in this file?

File: appium-tests/helpers/wdHelper.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+'use strict';
+
+var wd = global.WD || require('wd');
+var driver;
+
+module.exports.getDriver = function (platform, callback) {
+    var serverConfig = {
+        host: 'localhost',
+        port: 4723
```
Comment (dblotsky) [inline appium-tests/helpers/wdHelper.js]: Please factor these out into constants.

File: appium-tests/ios/ios.spec.js (hunk 1 above)
Comment (alsorokin) [inline appium-tests/ios/ios.spec.js]: At this point, the failure already has been loudly announced, this logic is just to skip current beforeEach function.

File: appium-tests/android/android.spec.js (hunk 4 above)
Comment (alsorokin) [inline appium-tests/android/android.spec.js]: Yeah, but it looks for the item in an array, not for a substring in a string.
I guess I'll just have to rename this one.

File: appium-tests/android/android.spec.js (hunk 5 above)
Comment (alsorokin) [inline appium-tests/android/android.spec.js]: There is an expect only if the driver has failed to create the session.

File: appium-tests/android/android.spec.js (hunk 7 above)
Comment (alsorokin) [inline appium-tests/android/android.spec.js]: This is just a filler for the phone. it's not used anywhere else.

File: appium-tests/helpers/contactsHelper.js (hunk 9 above)
Comment (alsorokin) [inline appium-tests/helpers/contactsHelper.js]: This is the code to run on the client side. We inject it by calling `driver.execute()` from the specs.

File: appium-tests/android/android.spec.js (hunk 3 above)
Comment (alsorokin) [inline appium-tests/android/android.spec.js]: I was not aware that they did implement it. Such a nice function, gonna use it a lot.

File: appium-tests/helpers/contactsHelper.js (hunk 9 above)
Comment (dblotsky) [inline appium-tests/helpers/contactsHelper.js]: I'm very much against this pattern. Can you describe in more detail why this is necessary, and if there are other ways to implement these tests?

Comment (alsorokin) [inline appium-tests/helpers/contactsHelper.js]: There are other way: we could just click the buttons in the mobilespec app's manual tests.
I've found a couple of downsides of such approach:
- We are tightly tied to the mobilespec app and its implementation: should there be changes to it's markup/navigation, or to the manual tests, we're bound to change the appium tests too
- Appium tests look more complicated. For example, to debug a failing Appium test which clicks a button in the manual test I need to look at the Appium code, see which button is clicked, like here: `elementByXPath('//a[text()="Add new nameless contact"]')`, go to the manual tests, find the code that creates this particular button and finally find a click handler function for this particular button. All this opposed to just having the executed code close at hand, in the helper function.
- Clicking buttons can be difficult :) For example, Appium is having a huge problem scrolling the page to find a button which is out of view on iOS simulator.

Could you please describe your thoughts about the current implementation and why are you against it?

**Comment by alsorokin** [comment]
I'm going to merge it in soon. @dblotsky @riknoll @nikhilkh

File: appium-tests/helpers/contactsHelper.js (hunk 9 above)
Comment (dblotsky) [inline appium-tests/helpers/contactsHelper.js]: Currently the tests already depend on the Mobilespec app. This code should live in the app.

There are many reasons why writing code in strings is a bad pattern. Here are just a few:
- it's very hard to de-duplicate and refactor
- it's not checked for syntax, therefore:
    - it's easier to make mistakes in it
    - it's more difficult to fix bugs in it
    - even if it looks fine, it can break at runtime because errors can be inserted at runtime
- it's difficult to edit and read because:
    - it's not highlighted
    - its format is hard to see

In short, it's a very dangerous pattern and we can easily avoid it, so we should definitely not be using it. I agree with you that our UI code lives too far from our UI tests, but I think that's a problem with Mobilespec's design. There are many ways to fix it (e.g. having separate test apps that live with each plugin), and we should discuss fixing Mobilespec as a separate task.

Comment (sgrebnov) [inline appium-tests/helpers/contactsHelper.js]: writing code in strings is a bad pattern indeed. But I don't like the idea to rely on buttons/divs from manual tests. 

I propose the following:
1. Step1. Add helper function to appium which is based on execute, but takes single PARAMETERLESS function as argument repreneting function to be executed on app side. So you will be able to do `return execute(fn.toString())` ti run js code on app side, for example
```
.execute(function() {
    navigator.contacts.create('someParams', winCallback, falseCallback)
})
```
2. Step2. Inject gloabal win and fail functions to be used on app context as a first step before running tests, so you will be able to pass them in all functions, for example `__appiumWinCallback and __appiumFailCallback`

So you will be able to run the following code
```
.execute(function() {
    navigator.contacts.create('someParams', __appiumWinCallback, __appiumFailCallback)
})
```
3. Step2. Inject extra function to return  last execution results (should be based on __appiumWinCallback and __appiumFailCallback implementations), sample usage (appium side)

```
var result = execute('_appiumLastTestResult')
result.win <- raw result from __appiumWinCallback callback (if called)
result.fail <- raw result from __appiumFailCallback fail 

```

Comment (sgrebnov) [inline appium-tests/helpers/contactsHelper.js]: Looks like there is also executeAsync which could be used to notify when win or fail callback is called: http://webdriver.io/api/protocol/executeAsync.html

**Comment by dblotsky** [comment]
@sgrebnov `fn.toString` is also a pretty nasty pattern. Adding buttons for manual tests is literally what manual tests are. We control both the tests and the UI. Why is that a bad thing? This is one of those cases where applying the DRY (Don't Repeat Yourself) principle doesn't make things much better and makes the code a lot harder to maintain. Just modify the test app. That way we can also test it manually without Appium.

**Comment by sgrebnov** [comment]
@dblotsky  We don't have manual tests - we have buttons demonstrating api calls (for both api classes - with and without user interaction), there are no validation logic there.

There are two options how we can proceed IMO (i'm fine with both of them):
1. Made them as real tests (as you mentioned above so you can run them w/o appium). I actually support idea to be able to run tests manually if needed. Where do you think should be validation logic in this case (appium side or client app side)? Should we create separate button for each option/combination for Camera tests for example, or use switches/etc. Also manual api code writes  results to divs/dom elements so we will have to bring extra complexity to tests to find/parse results from dom elements before validation.
2. Don't use those buttons at all (same reason why unit tests don't rely on calling buttons) - calling single cordova function from appium and getting callback result is simple. This may lead to more clean/readable code and tests definitions and simplify tests maintenance.

Agree regarding `fn.toString()`, but there is a special function for this - you can execute function in browser context from appium.
http://webdriver.io/api/protocol/execute.html

**Comment by dblotsky** [comment]
@sgrebnov pure code tests that don't need UI interaction should not be in Appium tests. Pure code tests that need UI interaction (like taking a picture) should have a button for them. Option 1 is the path we should take: the app should be automatically and manually testable. Validation logic should follow what humans do, which is to say that it should live in Appium and should inspect output values on the UI.

**Comment by nikhilkh** [comment]
@riknoll @dblotsky Can you please review again? This seems to be converging to a good design now.

File: appium-tests/common/common.spec.js [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+                    expect(contact.name.familyName).toBe('McContact');
+                })
+                .then(function () {
+                    return pickContact('Urist McContact');
+                })
+                .then(function (contact) {
+                    expect(contact.name.givenName).toBe('Urist');
+                    expect(contact.name.familyName).toBe('McContact');
+                })
+                .fail(saveScreenshotAndFail)
```
Comment (nikhilkh) [inline appium-tests/common/common.spec.js]: Instead of doing this - you could add a jasmine thing to ensure that on every failure we get screenshot:

https://github.com/webdriverio/webdriverio/issues/556

File: appium-tests/common/common.spec.js [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+                .finally(done);
```
Comment (nikhilkh) [inline appium-tests/common/common.spec.js]: You want to end the promise chain with .done() - otherwise failures can get lost.

File: appium-tests/common/common.spec.js [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+                    for (var i = 0; i < contexts.length ; i++) {
+                        if (contexts[i].indexOf('WEBVIEW') >= 0) {
+                            webviewContext = contexts[i];
+                            found = true;
+                        }
+                    }
+                    if (!found) {
+                        // no webview context, the app is still loading
+                        return driver
+                            .sleep(10000)
```
Comment (dblotsky) [inline appium-tests/common/common.spec.js]: Please make this a constant.

File: appium-tests/common/common.spec.js [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+                var cbEl = document.createElement('div');
+                cbEl.id = cbId;
+                cbEl.style.display = 'none';
+                navigator.contacts.pickContact(function (contact) {
+                    cbEl.innerHTML = JSON.stringify(contact);
+                    document.body.appendChild(cbEl);
+                }, function (err) {
+                    cbEl.innerHTML = 'ERROR: ' + err;
+                    document.body.appendChild(cbEl);
+                });
```
Comment (dblotsky) [inline appium-tests/common/common.spec.js]: @riknoll and I thought about this approach and realized that this can be done in a cleaner way by using promises. Instead of creating a DOM element to store the result, you can wrap the `pickContact` call in a promise and save the promise on the `navigator` object. Then you can just call `.done()` on it using `execAsync` instead of using `waitForElementById`.

File: appium-tests/common/common.spec.js (hunk 11 above)
Comment (alsorokin) [inline appium-tests/common/common.spec.js]: Taking screenshot should be a part of a promise chain. In fact, some of our current failures (camera plugin) on the CI are there because in that old version of camera tests we take screenshots without respect to the promise chain:
https://github.com/apache/cordova-plugin-camera/blob/master/appium-tests/android/android.spec.js#L62

**Comment by alsorokin** [comment]
Related PR to Medic: https://github.com/apache/cordova-medic/pull/85
